src/main/scala/com/wavesplatform/state/Diff.scala

```diff
@@ -71,14 +76,46 @@ object AccountDataInfo {
   }
 }
 
+sealed abstract class Sponsorship
+case class SponsorshipValue(minFee: Long) extends Sponsorship
+case object SponsorshipNoInfo             extends Sponsorship
+
+object Sponsorship {
+  val FeeUnit = 100000
+
+  implicit val sponsorshipMonoid: Monoid[Sponsorship] = new Monoid[Sponsorship] {
+    override def empty: Sponsorship = SponsorshipNoInfo
+
+    override def combine(x: Sponsorship, y: Sponsorship): Sponsorship = y match {
+      case SponsorshipNoInfo => x
+      case _                 => y
+    }
+  }
+
+  def sponsoredFeesSwitchHeight(blockchain: Blockchain, fs: FunctionalitySettings): Int =
+    blockchain
+      .featureActivationHeight(BlockchainFeatures.FeeSponsorship.id)
+      .map(h => h + fs.blocksForFeatureActivation(h))
+      .getOrElse(Int.MaxValue)
+
+  def toWaves(assetFee: Long, sponsorship: Long): Long = {
+    val waves = (BigDecimal(assetFee) * BigDecimal(Sponsorship.FeeUnit)) / BigDecimal(sponsorship)
+    if (waves > Long.MaxValue) {
+      throw new java.lang.ArithmeticException("Overflow")
+    }
+    waves.toLong
+  }
+}
+
 case class Diff(transactions: Map[ByteStr, (Int, Transaction, Set[Address])],
                 portfolios: Map[Address, Portfolio],
                 issuedAssets: Map[AssetId, AssetInfo],
                 aliases: Map[Alias, Address],
                 orderFills: Map[ByteStr, VolumeAndFee],
                 leaseState: Map[ByteStr, Boolean],
                 scripts: Map[Address, Option[Script]],
-                accountData: Map[Address, AccountDataInfo]) {
+                accountData: Map[Address, AccountDataInfo],
+                sponsorship: Map[AssetId, Sponsorship]) {
```

Comment:
We already have `issuedAssets: Map[AssetId, AssetInfo]` why can't `Sponsorship be part of `AssetInfo` ?